src/Talis/Persona/Client/Login.php

```diff
@@ -138,8 +138,9 @@ public function getPersistentId()
             // the prefix of the persona profile
             foreach ($_SESSION[self::LOGIN_PREFIX . ':loginSSO']['gupid'] as $gupid) {
                 $loginProvider = $_SESSION[self::LOGIN_PREFIX . ':loginProvider'];
```

Comment:
Just out of interest... a thought experiment.

Say we have a Persona provider called `tenant`, which contains IDP steps `tenant-1` and `tenant-2`.

The calling application should know about the Persona provider name - `tenant` in this case.  However, `tenant` is never the GUPID namespace, because it's not an IDP shortcode.  The calling application should not know anything about the individual steps - that is entirely handled by Persona.  In other words, the calling app knows nothing about the IDP shortcodes.

How is the calling app supposed to be able to select a GUPID, if it doesn't know the namespace used to identify that GUPID?  Can this function ever work correctly?

It strikes me that the only reliable way to determine this is have Persona respond with the currently activated GUPID and persistent ID, rather than relying on the client to derive it. (Moreover, having Persona return it centralises the decision - otherwise, all the apps and clients need to reimplement it, which is less efficient and introduces the risk of bugs.)

Reply:
I've just discussed this in depth with @kiyanwang and it sounds like he had the same concerns, and that we **are not using this method for TARL to get the active persistent ID**.  Instead, that is being obtained in a valid Persona-led approach, delivered in the server-side login payload.

*phew*.

I can confirm that from querying the databases that we do not have any offending GUPIDs for a tenancy that would not have a prefix match in its GUPIDs.

I think we should strongly consider removing this method, though, since it's likely doing the wrong thing, if we do use it anywhere.